### PR TITLE
Add NioGroup for use in different transports

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit is related to #27260. It adds a base `NioGroup` for use in
different transports. This class creates and starts the underlying
selectors. Different protocols or transports are established by passing
the `ChannelFactory` to the `bindServerChannel` or `openChannel`
methods. This allows a `TcpChannelFactory` to be passed which will
create and register channels that support the elasticsearch tcp binary
protocol or a channel factory that will create http channels (or other).

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                threadFactory.newThread(acceptor::runLoop).start();
+                acceptor.isRunningFuture().actionGet();
+            }
+        }
+    }
+
+    private void shutdownSelector(ESSelector selector) {
+        try {
+            selector.close();
+        } catch (Exception e) {
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java]: it might make sense to deduplicate this exception otherwise we might spam logs?

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    public void start() throws IOException {
+        for (int i = 0; i < socketSelectorCount; ++i) {
+            SocketSelector selector = new SocketSelector(socketEventHandlerFunction.apply(logger));
+            socketSelectors.add(selector);
+        }
+
+        startSelectors(socketSelectors, socketSelectorThreadFactory);
+
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java]: extra newlines

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    private final ThreadFactory acceptorThreadFactory;
+    private final ArrayList<AcceptingSelector> acceptors;
+    private final BiFunction<Logger, Supplier<SocketSelector>, AcceptorEventHandler> acceptorEventHandlerFunction;
+    private RoundRobinSupplier<AcceptingSelector> acceptorSupplier;
+
+    private final int socketSelectorCount;
+    private final ThreadFactory socketSelectorThreadFactory;
+    private final ArrayList<SocketSelector> socketSelectors;
+    private final Function<Logger, SocketEventHandler> socketEventHandlerFunction;
+    private RoundRobinSupplier<SocketSelector> socketSelectorSupplier;
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java]: can be final?

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+public class NioGroup {
+
+    private final Logger logger;
+
+    private final int acceptorCount;
+    private final ThreadFactory acceptorThreadFactory;
+    private final ArrayList<AcceptingSelector> acceptors;
+    private final BiFunction<Logger, Supplier<SocketSelector>, AcceptorEventHandler> acceptorEventHandlerFunction;
+    private RoundRobinSupplier<AcceptingSelector> acceptorSupplier;
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java]: can be final?

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+import org.elasticsearch.transport.nio.channel.NioSocketChannel;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.concurrent.ThreadFactory;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java]: can you summarize what this does?

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        }
+        acceptors.clear();
+
+        for (SocketSelector selector : socketSelectors) {
+            shutdownSelector(selector);
+        }
+        socketSelectors.clear();
+    }
+
+    private static <S extends ESSelector> void startSelectors(Iterable<S> selectors, ThreadFactory threadFactory) {
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java]: if we fail to start do we close the started ones

**Review by s1monw — CHANGES_REQUESTED** [review]
I left some comments

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java (hunk 4 above)
Comment (tbrooks8) [reply test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java]: I don't think they can be final in the current design. Creating a `Selector` can throw an exception which is why I do it in `start()`.

I guess the options are:
1. I could make them `volatile`
2. I could throw `IOException` from the ctor. Although with handling exceptions and closing in the ctor.

Thoughts?

**Comment by tbrooks8** [comment]
@s1monw - based on your comment about making the suppliers final, I decided to put in some work relating to ensuring that the state of the `NioGroup` is consistent between start and close. With those changes, this is ready for another review.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java (hunk 4 above)
Comment (s1monw) [reply test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java]: I am leaning towards 2. it's much cleaner

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        this.acceptorEventHandlerFunction = acceptorEventHandlerFunction;
+        this.socketSelectorThreadFactory = socketSelectorThreadFactory;
+        this.acceptorCount = acceptorCount;
+        this.socketSelectorCount = socketSelectorCount;
+        this.socketEventHandlerFunction = socketEventHandlerFunction;
+        acceptors = new ArrayList<>(this.acceptorCount);
+        socketSelectors = new ArrayList<>(this.socketSelectorCount);
+    }
+
+    public synchronized void start() throws IOException {
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java]: if we can prevent this kind of state handling we should. My experience with stuff like this is consistently negative so lets try to move it to the ctor and make it clear once the group is created it's ready to use. I still think we should have an `ensureOpen()` method that checks a boolean if we are still open.

**Review by s1monw — CHANGES_REQUESTED** [review]

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java [hunk 8]
```diff
@@ -116,27 +89,27 @@ public synchronized void start() throws IOException {
...
         acceptorSupplier = new RoundRobinSupplier<>(acceptors.toArray(new AcceptingSelector[acceptors.size()]));
-        state = RUNNING;
     }
 
     public <S extends NioServerSocketChannel> S bindServerChannel(InetSocketAddress address, ChannelFactory<S, ?> factory)
         throws IOException {
-        ensureRunning();
-        if (acceptorCount == 0) {
+        ensureOpen();
+        if (acceptors.size() == 0) {
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java]: nit: maybe us isEmpty()?

**Review by s1monw — APPROVED** [review]
LGTM
